File: cmd/desktop/internal/cache/cache.go
package cache

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"time"

	_ "modernc.org/sqlite"
)

const maxEntries = 50

// Entry represents a cached notification.
type Entry struct {
	ID         string
	Title      string
	Body       string
	URL        string
	Channel    string
	Source     string
	ReceivedAt time.Time
	Unread     bool
	AckedAt    *time.Time
	ExpiresAt  *time.Time
}

// Cache manages local notification history in SQLite.
type Cache struct {
	db *sql.DB
}

// New creates a new Cache, initializing the SQLite database.
func New() (*Cache, error) {
	dbPath, err := dbFilePath()
	if err != nil {
		return nil, err
	}

	dir := filepath.Dir(dbPath)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return nil, fmt.Errorf("create cache dir: %w", err)
	}

	db, err := sql.Open("sqlite", dbPath)
	if err != nil {
		return nil, fmt.Errorf("open cache db: %w", err)
	}

	// Create table
	_, err = db.Exec(`CREATE TABLE IF NOT EXISTS notifications (
		id TEXT PRIMARY KEY,
		title TEXT NOT NULL,
		body TEXT,
		url TEXT,
		channel TEXT,
		source TEXT,
		received_at DATETIME DEFAULT CURRENT_TIMESTAMP,
		unread INTEGER DEFAULT 1,
		acked_at DATETIME,
		expires_at DATETIME
	)`)
	if err != nil {
		db.Close()
		return nil, fmt.Errorf("create table: %w", err)
	}

	// Migrate: add lifecycle columns if they don't exist (for existing databases)
	db.Exec(`ALTER TABLE notifications ADD COLUMN unread INTEGER DEFAULT 1`)
	db.Exec(`ALTER TABLE notifications ADD COLUMN acked_at DATETIME`)
	db.Exec(`ALTER TABLE notifications ADD COLUMN expires_at DATETIME`)

	return &Cache{db: db}, nil
}

// Add stores a notification entry, enforcing the max limit.
func (c *Cache) Add(e Entry) error {
	_, err := c.db.Exec(`INSERT OR REPLACE INTO notifications (id, title, body, url, channel, source, received_at, unread, acked_at, expires_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		e.ID, e.Title, e.Body, e.URL, e.Channel, e.Source, e.ReceivedAt, boolToInt(e.Unread), e.AckedAt, e.ExpiresAt)
	if err != nil {
		return err
	}

	// Enforce max entries
	_, err = c.db.Exec(`DELETE FROM notifications WHERE id NOT IN (SELECT id FROM notifications ORDER BY received_at DESC LIMIT ?)`, maxEntries)
	return err
}

// MarkRead marks a single notification as read.
func (c *Cache) MarkRead(id string) error {
	_, err := c.db.Exec(`UPDATE notifications SET unread = 0 WHERE id = ?`, id)
	return err
}

// MarkAllRead marks all unread notifications as read.
func (c *Cache) MarkAllRead() error {
	_, err := c.db.Exec(`UPDATE notifications SET unread = 0 WHERE unread = 1`)
	return err
}

// UnreadCount returns the number of unexpired unread notifications.
func (c *Cache) UnreadCount() (int, error) {
	var count int
	err := c.db.QueryRow(`SELECT COUNT(*) FROM notifications WHERE unread = 1 AND (expires_at IS NULL OR expires_at > ?)`, time.Now()).Scan(&count)
	return count, err
}

// MarkAckSynced marks a notification as read and records the ack timestamp.
func (c *Cache) MarkAckSynced(id string) error {
	now := time.Now()
	_, err := c.db.Exec(`UPDATE notifications SET unread = 0, acked_at = ? WHERE id = ?`, now, id)
	return err
}

// GetUnexpiredUnread returns all unread notifications that have not expired.
func (c *Cache) GetUnexpiredUnread() ([]Entry, error) {
	rows, err := c.db.Query(`SELECT id, title, body, url, channel, source, received_at, unread, acked_at, expires_at FROM notifications WHERE unread = 1 AND (expires_at IS NULL OR expires_at > ?) ORDER BY received_at DESC`, time.Now())
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	return scanEntries(rows)
}

// Recent returns the N most recent notifications.
func (c *Cache) Recent(n int) ([]Entry, error) {
	rows, err := c.db.Query(`SELECT id, title, body, url, channel, source, received_at, unread, acked_at, expires_at FROM notifications ORDER BY received_at DESC LIMIT ?`, n)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	return scanEntries(rows)
}

// Close closes the database connection.
func (c *Cache) Close() error {
	if c.db != nil {
		return c.db.Close()
	}
	return nil
}

func scanEntries(rows *sql.Rows) ([]Entry, error) {
	var entries []Entry
	for rows.Next() {
		var e Entry
		var body, url, channel, source sql.NullString
		var unread int
		var ackedAt, expiresAt sql.NullTime
		if err := rows.Scan(&e.ID, &e.Title, &body, &url, &channel, &source, &e.ReceivedAt, &unread, &ackedAt, &expiresAt); err != nil {
			return nil, err
		}
		if body.Valid {
			e.Body = body.String
		}
		if url.Valid {
			e.URL = url.String
		}
		if channel.Valid {
			e.Channel = channel.String
		}
		if source.Valid {
			e.Source = source.String
		}
		e.Unread = unread == 1
		if ackedAt.Valid {
			t := ackedAt.Time
			e.AckedAt = &t
		}
		if expiresAt.Valid {
			t := expiresAt.Time
			e.ExpiresAt = &t
		}
		entries = append(entries, e)
	}
	return entries, nil
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

func dbFilePath() (string, error) {
	switch runtime.GOOS {
	case "windows":
		localAppData := os.Getenv("LOCALAPPDATA")
		if localAppData == "" {
			return "", fmt.Errorf("LOCALAPPDATA not set")
		}
		return filepath.Join(localAppData, "overseer-desktop", "notifications.db"), nil
	default:
		home, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}
		return filepath.Join(home, ".config", "overseer-desktop", "notifications.db"), nil
	}
}

File: desktop/src-tauri/src/cache.rs
use chrono::{DateTime, Utc};
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::sync::Mutex;

const MAX_ENTRIES: i64 = 50;

/// A cached notification entry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entry {
    pub id: String,
    pub title: String,
    pub body: String,
    pub url: String,
    pub channel: String,
    pub source: String,
    pub received_at: DateTime<Utc>,
    pub unread: bool,
    pub acked_at: Option<DateTime<Utc>>,
    pub expires_at: Option<DateTime<Utc>>,
}

/// Notification cache backed by SQLite.
pub struct Cache {
    conn: Mutex<Connection>,
}

impl Cache {
    /// Create a new cache, initializing the SQLite database.
    pub fn new() -> Result<Self, String> {
        let db_path = Self::db_file_path()?;
        if let Some(dir) = db_path.parent() {
            std::fs::create_dir_all(dir).map_err(|e| format!("create cache dir: {}", e))?;
        }

        let conn =
            Connection::open(&db_path).map_err(|e| format!("open cache db: {}", e))?;

        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS notifications (
                id TEXT PRIMARY KEY,
                title TEXT NOT NULL,
                body TEXT,
                url TEXT,
                channel TEXT,
                source TEXT,
                received_at TEXT DEFAULT (datetime('now')),
                unread INTEGER DEFAULT 1,
                acked_at TEXT,
                expires_at TEXT
            );",
        )
        .map_err(|e| format!("create table: {}", e))?;

        Ok(Self {
            conn: Mutex::new(conn),
        })
    }

    /// Add a notification entry, enforcing the max limit.
    pub fn add(&self, entry: &Entry) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        conn.execute(
            "INSERT OR REPLACE INTO notifications (id, title, body, url, channel, source, received_at, unread, acked_at, expires_at) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10)",
            params![
                entry.id,
                entry.title,
                entry.body,
                entry.url,
                entry.channel,
                entry.source,
                entry.received_at.to_rfc3339(),
                entry.unread as i32,
                entry.acked_at.map(|t| t.to_rfc3339()),
                entry.expires_at.map(|t| t.to_rfc3339()),
            ],
        ).map_err(|e| format!("insert: {}", e))?;

        // Enforce max entries
        conn.execute(
            "DELETE FROM notifications WHERE id NOT IN (SELECT id FROM notifications ORDER BY received_at DESC LIMIT ?1)",
            params![MAX_ENTRIES],
        ).map_err(|e| format!("cleanup: {}", e))?;

        Ok(())
    }

    /// Mark a single notification as read.
    pub fn mark_read(&self, id: &str) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        conn.execute("UPDATE notifications SET unread = 0 WHERE id = ?1", params![id])
            .map_err(|e| format!("mark read: {}", e))?;
        Ok(())
    }

    /// Mark all notifications as read.
    pub fn mark_all_read(&self) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        conn.execute("UPDATE notifications SET unread = 0 WHERE unread = 1", [])
            .map_err(|e| format!("mark all read: {}", e))?;
        Ok(())
    }

    /// Get the count of unexpired unread notifications.
    pub fn unread_count(&self) -> Result<i32, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let now = Utc::now().to_rfc3339();
        let count: i32 = conn
            .query_row(
                "SELECT COUNT(*) FROM notifications WHERE unread = 1 AND (expires_at IS NULL OR expires_at > ?1)",
                params![now],
                |row| row.get(0),
            )
            .map_err(|e| format!("unread count: {}", e))?;
        Ok(count)
    }

    /// Mark a notification as ack-synced (read + acked_at timestamp).
    pub fn mark_ack_synced(&self, id: &str) -> Result<(), String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let now = Utc::now().to_rfc3339();
        conn.execute(
            "UPDATE notifications SET unread = 0, acked_at = ?1 WHERE id = ?2",
            params![now, id],
        )
        .map_err(|e| format!("ack sync: {}", e))?;
        Ok(())
    }

    /// Get the N most recent notifications.
    pub fn recent(&self, n: i32) -> Result<Vec<Entry>, String> {
        let conn = self.conn.lock().map_err(|e| e.to_string())?;
        let mut stmt = conn
            .prepare("SELECT id, title, body, url, channel, source, received_at, unread, acked_at, expires_at FROM notifications ORDER BY received_at DESC LIMIT ?1")
            .map_err(|e| format!("prepare: {}", e))?;

        let entries = stmt
            .query_map(params![n], |row| {
                Ok(Self::row_to_entry(row))
            })
            .map_err(|e| format!("query: {}", e))?
            .filter_map(|r| r.ok())
            .collect();

        Ok(entries)
    }

    fn row_to_entry(row: &rusqlite::Row) -> Entry {
        let received_at_str: String = row.get(6).unwrap_or_default();
        let unread_int: i32 = row.get(7).unwrap_or(1);
        let acked_at_str: Option<String> = row.get(8).unwrap_or(None);
        let expires_at_str: Option<String> = row.get(9).unwrap_or(None);

        Entry {
            id: row.get(0).unwrap_or_default(),
            title: row.get(1).unwrap_or_default(),
            body: row.get(2).unwrap_or_default(),
            url: row.get(3).unwrap_or_default(),
            channel: row.get(4).unwrap_or_default(),
            source: row.get(5).unwrap_or_default(),
            received_at: DateTime::parse_from_rfc3339(&received_at_str)
                .map(|dt| dt.with_timezone(&Utc))
                .unwrap_or_else(|_| Utc::now()),
            unread: unread_int == 1,
            acked_at: acked_at_str
                .and_then(|s| DateTime::parse_from_rfc3339(&s).ok())
                .map(|dt| dt.with_timezone(&Utc)),
            expires_at: expires_at_str
                .and_then(|s| DateTime::parse_from_rfc3339(&s).ok())
                .map(|dt| dt.with_timezone(&Utc)),
        }
    }

    fn db_file_path() -> Result<PathBuf, String> {
        dirs::config_dir()
            .map(|d| d.join("overseer-desktop").join("notifications.db"))
            .ok_or_else(|| "cannot determine config dir".to_string())
    }
}
